### PR TITLE
[addons][base][filesystem][gui] improvement and fixes.

### PR DESCRIPTION
**Pull request by AlwinEsch**
## Description

Commit 1:
-----------------
**[addons][filesystem] add function to get available disc space**

This add a function to ask about available and total disc space.
For cases where e.g. inputstream addon use a file storage for timeshift in livestream

Commit 2:
-----------------
**[addons] fix and improve instance creation calls**

Before was there some not good parts:
- To use "throw" inside "C" ABI on other side is not correct
- Fix mem leak if CreateInstance returned other value as ADDON_STATUS_OK
  - If other was it not stored in Kodi and the DestroyInstance not called
  - This stores now in all cases if present
- Make CreateInstance with ADDON_STATUS_NOT_IMPLEMENTED optional
  - Thought for case where a addon without instances is used
  - If e.g. a new addon type is added for background works (and not Kodi call itself)

Seen this on PVR rework where by CreateInstance in some cases another return values
as ADDON_STATUS_OK given.

Commit 3:
-----------------
**[addons][gui] fix crash by wrong shared_from_this call**

There it was set before as temporary stack value where it was not available
as shared_ptr.

This use the make_shared about and the crash is fixed.

Crash Log:
```
    at /usr/include/c++/9/bits/shared_ptr_base.h:900
    (this=0x7fffffff9dc0, __r=...) at /usr/include/c++/9/bits/shared_ptr_base.h:1193
    (this=0x7fffffff9dc0, __r=std::weak_ptr<class ADDON::IAddon> (empty) = {...}) at /usr/include/c++/9/bits/shared_ptr.h:276
    (kodiBase=0x7fff4c023760, xml_filename=0x55555b3c9350 "DialogRecordSettings.xml", default_skin=0x7fffffffabc0 "skin.fallback", as_dialog=false, is_media=true)
    at /home/alwin/Development/Kodi/kodi-Matrix/xbmc/addons/interfaces/gui/Window.cpp:159
```

Commit 4:
-----------------
**[addons] use addon ID instead of Name for Log**

Before was the addon name written in Kodi log, this can be a bit
confusing as it is more easy to identify by ID as by name.

Commit 5:
-----------------
**[addons][filesystem] use const on CFile get calls**

Some addons takes them with this inside a parent protected class entries.
This allow the read then.

*Also API version increase inside them*

Commit 6:
-----------------
**[addons][base] allow optional set of default value on GetSetting... calls**

This to prevent on addon a extra "if (!...)" to set the default if not found.

*Also API version increase inside them*

## Motivation and Context
All older compiled addons are still operational after this change.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/addons/interfaces/Filesystem.cpp [hunk 1]
```diff
@@ -426,6 +428,35 @@ char* Interface_Filesystem::translate_special_protocol(void* kodiBase, const cha
...
+        LOGERROR,
+        "Interface_Filesystem::{} - invalid data (addon='{}', path='{}, capacity='{}, free='{}, "
+        "available='{})",
+        __FUNCTION__, kodiBase, static_cast<const void*>(path), static_cast<void*>(capacity),
+        static_cast<void*>(free), static_cast<void*>(available));
+    return false;
+  }
+
+  std::error_code ec;
+  auto freeSpace = space(CSpecialProtocol::TranslatePath(path), ec);
```
Comment (garbear) [inline xbmc/addons/interfaces/Filesystem.cpp:450]: Where is `space()` defined?

**Review by garbear — APPROVED** [review]

File: xbmc/addons/interfaces/Filesystem.cpp (hunk 1 above)
Comment (AlwinEsch) [reply xbmc/addons/interfaces/Filesystem.cpp:450]: Is here, also searched a while to find:
- https://github.com/xbmc/xbmc/blob/master/xbmc/platform/Filesystem.h#L25
- https://github.com/xbmc/xbmc/blob/master/xbmc/platform/posix/Filesystem.cpp#L34
- https://github.com/xbmc/xbmc/blob/master/xbmc/platform/win32/Filesystem.cpp#L23

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h [hunk 2]
```diff
@@ -1254,6 +1254,50 @@ inline std::string TranslateSpecialProtocol(const std::string& source)
...
+///
+/// **Example:**
+/// ~~~~~~~~~~~~~{.cpp}
+/// #include <kodi/Filesystem.h>
+/// ...
+/// std::string path = "special://temp";
+/// uint64_t capacity = -1;
+/// uint64_t free = -1;
+/// uint64_t available = -1;
+/// kodi::vfs::GetDiscSpace(path, capacity, free, available);
```
Comment (phunkyfish) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h:1282]: Typo: GetDiskSpace

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h [hunk 3]
```diff
@@ -1254,6 +1254,50 @@ inline std::string TranslateSpecialProtocol(const std::string& source)
...
+/// fprintf(stderr, "Path '%s' sizes:\n", path.c_str());
+/// fprintf(stderr, " - capacity:  %lu MByte\n", capacity / 1024 / 1024);
+/// fprintf(stderr, " - free:      %lu MByte\n", free / 1024 / 1024);
+/// fprintf(stderr, " - available: %lu MByte\n", available / 1024 / 1024);
+/// ~~~~~~~~~~~~~
+///
+inline bool GetDiscSpace(const std::string& path,
```
Comment (phunkyfish) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/Filesystem.h:1289]: Typo: GetDiskSpace

**Review by phunkyfish — CHANGES_REQUESTED** [review]

**Comment by phunkyfish** [comment]
Tested this morning. It looks like the space() call only works for a drive, i.e. "/" on MacOS.

So what we should probably do is translate the path and pass either "/" or the drive letter to Kodi's `GetDiskSpace()` function.

**Comment by phunkyfish** [comment]
Correction. This does work correctly. I had a typo in the special path I used for testing.

**Review by phunkyfish — APPROVED** [review]
Nice, also tested other commits on inputstream and PVR and all works ok.

**Comment by AlwinEsch** [comment]
Tested also now on Windows with and was OK:
```cpp
std::string path = "special://temp"; // OK
uint64_t capacity = -1;
uint64_t free = -1;
uint64_t available = -1;
kodi::vfs::GetDiscSpace(path, capacity, free, available);
fprintf(stderr, "Path '%s' sizes:\n", path.c_str());
fprintf(stderr, " - capacity:  %lu MByte\n", capacity / 1024 / 1024);
fprintf(stderr, " - free:      %lu MByte\n", free / 1024 / 1024);
fprintf(stderr, " - available: %lu MByte\n", available / 1024 / 1024);

path = "D:\\"; // OK
capacity = -1;
free = -1;
available = -1;
kodi::vfs::GetDiscSpace(path, capacity, free, available);
fprintf(stderr, "Path '%s' sizes:\n", path.c_str());
fprintf(stderr, " - capacity:  %lu MByte\n", capacity / 1024 / 1024);
fprintf(stderr, " - free:      %lu MByte\n", free / 1024 / 1024);
fprintf(stderr, " - available: %lu MByte\n", available / 1024 / 1024);

path = "smb://alwin-ms-7b47/OSTest"; // As expected not work
capacity = -1;
free = -1;
available = -1;
kodi::vfs::GetDiscSpace(path, capacity, free, available);
fprintf(stderr, "Path '%s' sizes:\n", path.c_str());
fprintf(stderr, " - capacity:  %lu MByte\n", capacity / 1024 / 1024);
fprintf(stderr, " - free:      %lu MByte\n", free / 1024 / 1024);
fprintf(stderr, " - available: %lu MByte\n", available / 1024 / 1024);

path = "Z:\\"; // OK (is to a network drive)
capacity = -1;
free = -1;
available = -1;
kodi::vfs::GetDiscSpace(path, capacity, free, available);
fprintf(stderr, "Path '%s' sizes:\n", path.c_str());
fprintf(stderr, " - capacity:  %lu MByte\n", capacity / 1024 / 1024);
fprintf(stderr, " - free:      %lu MByte\n", free / 1024 / 1024);
fprintf(stderr, " - available: %lu MByte\n", available / 1024 / 1024);

path = "\\\\alwin-ms-7b47\\OSTest"; // OK
capacity = -1;
free = -1;
available = -1;
kodi::vfs::GetDiscSpace(path, capacity, free, available);
fprintf(stderr, "Path '%s' sizes:\n", path.c_str());
fprintf(stderr, " - capacity:  %lu MByte\n", capacity / 1024 / 1024);
fprintf(stderr, " - free:      %lu MByte\n", free / 1024 / 1024);
fprintf(stderr, " - available: %lu MByte\n", available / 1024 / 1024);
```

```
Path 'special://temp' sizes:
 - capacity:  121212 MByte
 - free:      3185 MByte
 - available: 3185 MByte
Path 'D:\' sizes:
 - capacity:  1907161 MByte
 - free:      1778017 MByte
 - available: 1778017 MByte
Path 'smb://alwin-ms-7b47/OSTest' sizes:
 - capacity:  4294967295 MByte
 - free:      4294967295 MByte
 - available: 4294967295 MByte
Path 'Z:\' sizes:
 - capacity:  912672 MByte
 - free:      54033 MByte
 - available: 54033 MByte
Path '\\alwin-ms-7b47\OSTest' sizes:
 - capacity:  912672 MByte
 - free:      54033 MByte
 - available: 54033 MByte
```

**Comment by phunkyfish** [comment]
Did the failing case return false?

**Comment by AlwinEsch** [comment]
Have rechecked, yes on Windows and Linux it have the false then.

**Comment by phunkyfish** [comment]
Perfect, nice work 👍

**Comment by AlwinEsch** [comment]
jenkins build this with addons please

**Pull request merged**